File: server/src/game/effects/DamageEffect.ts
import type { Minion } from '../Minion.js';
import { Player } from '../Player.js';
import type { ICardEffect } from './ICardEffect.js';
import type { EffectContext } from './EffectContext.js';

export class DamageEffect implements ICardEffect {
  constructor(public readonly amount: number) {}

  execute(context: EffectContext): void {
    const target = context.target;

    if (target instanceof Player) {
      const actual = target.heroState.takeDamage(this.amount);
      if (target.id === context.opponent.id) {
        context.player.recordDamage(actual);
      }
      return;
    }

    if (target) {
      const minion = target as Minion;
      const actual = minion.takeDamage(this.amount);
      if (minion.ownerId === context.opponent.id) {
        context.player.recordDamage(actual);
      }
      return;
    }
  }
}

File: server-java/src/main/java/vn/coincard/server/game/effects/DamageEffect.java
package vn.coincard.server.game.effects;

import vn.coincard.server.game.Minion;
import vn.coincard.server.game.Hero;

/** Records actual damage only (no overkill), enemy side only. */
public class DamageEffect implements CardEffect {
  private final int amount;

  public DamageEffect(int amount) { this.amount = amount; }

  @Override
  public void execute(EffectContext context) {
    if (context.target instanceof Hero hero) {
      int actual = hero.takeDamage(amount);
      if (hero == context.opponent.heroState()) context.player.recordDamage(actual);
      return;
    }
    if (context.target instanceof Minion m) {
      int actual = m.takeDamage(amount);
      if (m.getOwnerId().equals(context.opponent.id())) context.player.recordDamage(actual);
    }
  }
}
